rbc/omniscidb.py

```diff
@@ -1148,3 +1243,157 @@ def compiler(self):
+
+        See RemoteJIT.caller_signature.__doc__.
+        """
+        if is_udtf(signature):
+            rtype = signature[0]
+            if not (rtype.is_int and rtype.bits == 32):
+                raise ValueError(
+                    f'UDTF implementation return type must be int32, got {rtype}')
+            rtypes = []
+            atypes = []
+            for atype in signature[1]:
+                if is_sizer(atype):
+                    sizer = get_sizer_enum(atype)
+                    if sizer not in user_specified_output_buffer_sizers:
+                        continue
+                    atype.annotation(sizer=sizer)
+                elif isinstance(atype, OmnisciTableFunctionManagerType):
+                    continue
+                elif isinstance(atype, OmnisciOutputColumnType):
+                    rtypes.append(atype.copy(OmnisciColumnType))
+                    continue
+                atypes.append(atype)
+            rtype = typesystem.Type(*rtypes, **dict(struct_is_tuple=True))
+            return rtype(*atypes, **signature._params)
+        return signature
+
+    def get_types(self, *values):
+        """Convert values to the corresponding typesystem types.
+
+        See RemoteJIT.get_types.__doc__.
+        """
+        types = []
+        for value in values:
+            if isinstance(value, RemoteCallCapsule):
+                typ = value.__typesystem_type__
+                if typ.is_struct and typ._params.get('struct_is_tuple'):
+                    types.extend(typ)
+                else:
+                    types.append(typ)
+            else:
+                types.append(typesystem.Type.fromvalue(value))
+        return tuple(types)
+
+    def normalize_function_type(self, ftype: typesystem.Type):
+        """Normalize typesystem function type.
+
+        See RemoteJIT.normalize_function_type.__doc__.
+        """
+        assert ftype.is_function, ftype
+        for atype in ftype[1]:
+            # convert `T foo` to `T | name=foo`
+            if 'name' not in atype.annotation() and atype.name is not None:
+                atype.annotation(name=atype.name)
+                atype._params.pop('name')
+        return ftype
+
+    def format_type(self, typ: typesystem.Type):
+        """Convert typesystem type to formatted string.
+
+        See RemoteJIT.format_type.__doc__.
+        """
+        if typ.is_function:
+            args = map(self.format_type, typ[1])
+            if is_udtf(typ):
+                return f'UDTF({", ".join(args)})'
+            else:
+                return f'({", ".join(args)}) -> {self.format_type(typ[0])}'
+        use_typename = False
+        if typ.is_struct and typ._params.get('struct_is_tuple'):
+            return f'({", ".join(map(self.format_type, typ))})'
+        if isinstance(typ, OmnisciOutputColumnType):
+            p = tuple(map(self.format_type, typ[0]))
+            typ = typesystem.Type(('OutputColumn',) + p, **typ._params)
```

Comment:
Shouldn't `typ` already be in the special type as an attribute?

Reply:
The aim of this remapping is to apply `format_type` to the parameters of OmnisciOutputColumnType (as T is a parameter in `OutputColumn<T>`). Since `OmnisciOutputColumnType` is a custom `typesystem.Type` then `typesystem` has no information about its internal structure (for instance, parameters of a custom type can be arbitrary Python objects). So, here we explicitly apply `format_type` to the parameters (line 1320) and construct a temporary custom type (line 1321) that does the right thing when calling its `tostring`.

Reply:
ok I am missing some pieces to fully get the whole logic 😅